Extract fatal notification into fatalError service, add support for EuiToast notifications

cjcenizal opened the pull request:
Relates to https://github.com/elastic/kibana/issues/15721. Partially addresses https://github.com/elastic/kibana/issues/8194. This simplifies the notification system (making it easier to integrate non-fatal notifications with EuiToasts) and adds support for displaying EuiToasts.

![image](https://user-images.githubusercontent.com/1238659/34283542-2a5febb6-e681-11e7-996a-3143277efa8f.png)

**To-do**

- [x] Make it easier to use toasts within functional tests, with `data-test-subj`, in a non-brittle way
- [x] Add Jest test coverage for `GlobalToastList`

## Functional test changes

### Avoid calling `clickToastOK()`

This method has traditionally been used as a hacky proxy for the successful creation of a dashboard, visualization, etc. Successful actions will trigger toasts instead of banner notifications, moving forward. So this method will become less and less useful and will eventually be removed.

### Saving dashboards and visualizations is self-confirming

By the same token, now when you call `saveDashboard()` or `saveVisualization()`, the method will wait until the confirmation toast appears and resolve successfully only once it appears. This means you no longer need to call these methods and then `clickToastOK()` afterwards.

### New `getBreadcrumbPageTitle()` method

The last breadcrumb behaves as a page title for many apps. This method will return the text within this breadcrumb, essentially returning the page title value.

### Make string-checking tests more robust

Visualization tests now leverage this method to confirm the name of a newly-created visualization. But they do so by asserting that the value _contains_ the name of the visualization, not that it matches it. A strict equality match is brittle and will result in a test failure if we decorate the page title with information in addition to the visualization name. But the UX is still intact so the test _should not_ fail. Checking that the page title contains the visualization name ensures that the test will still pass in this case. We should apply similar tactics where possible.

cjcenizal commented:
@Bargs Can you think of any reason not to remove the pattern-checker?

cjcenizal commented:
jenkins, test this

cjcenizal commented:
jenkins, test this

cjcenizal commented:
jenkins, test this

cjcenizal commented:
@nreese Could you take a look at my changes to Dashboard and let me know what you think? @thomasneirynck Could you do the same with Visualize? @bmcconaghy Would you mind doing the same with Management?

thomasneirynck commented:
jenkins, test this

cjcenizal commented:
jenkins, test this

cjcenizal commented:
jenkins test this

thomasneirynck commented on src/ui/public/courier/fetch/fetch_now.js:
```diff
@@ -30,7 +30,7 @@ export function FetchNowProvider(Private, Promise) {
       return req.retry();
     }))
-      .catch(courierNotifier.fatal);
+      .catch(error => fatalError(error, location));
```
seems like the optimizer is tripping over the uninitialized location variable

cjcenizal replied:
Oh shoot! Good catch, thanks. We should add a linting rule for that!

cjcenizal commented:
jenkins, test this

cjcenizal commented:
jenkins, test this

nreese commented on src/core_plugins/kibana/public/management/sections/indices/edit_index_pattern/scripted_field_editor/scripted_field_editor.js:
```diff
@@ -43,7 +43,11 @@ uiRoutes
         if (!this.field) {
-          notify.error(this.indexPattern + ' does not have a "' + fieldName + '" field.');
+          toastNotifications.add({
+            title: `"${this.indexPattern.title}" index pattern doesn't have a scripted field called "${fieldName}"`,
```
Do you want to use single quotes to be consistent with the confirm modals?

cjcenizal replied:
Yes! Great catch. Will fix.

cjcenizal commented:
@nreese I consider this level of skipped tests to be a blocker. I'm going to dig into them to identify the cause for their failure (I have a feeling it has something to do with the `clickToastOK` page object method). Once I've identified the cause, I'll fix several tests spread across several functional areas. Then I'll ask each functional team if they can submit PRs into this branch fixing the remaining tests, using my solution(s) as a guideline.

cjcenizal commented:
@elastic/kibana-sharing I've fixed most of the Dashboard tests. Only two are still skipped, and they both pertain to view-edit mode. I'm not sure what's causing the problem, but I'd be happy to debug them with you. Do you want to block this PR until we've fixed them or are you comfortable fixing them after we merge this?

@elastic/kibana-visualizations I skipped 5 Visualize tests. Two of them pertain to the spy toggle button, and the rest are TSVB-related. Can we work on these together?  Do you want to block this PR until we've fixed them or are you comfortable fixing them after we merge this?

nreese commented:
@cjcenizal let me take a look at the remaining dashboard tests this morning

cjcenizal commented:
I fixed all the tests. The only ones which are still skipped are the TSVB ones which are also failing on master. I've created https://github.com/elastic/kibana/issues/15977 to track this problem.

chrisronline commented on src/ui/public/notify/toasts/toast_notifications.js:
```diff
@@ -0,0 +1,20 @@
+    this.idCounter = 0;
+  }
+
+  add = toast => {
```
Just wondering, do you think it would be nice to expose an api more resembling the old notify one? Keep `add()`, but also include `success()`, `warning()`, etc? Those would just call add with the appropriate color.

cjcenizal replied:
Good suggestion guys, will do!

chrisronline commented on tasks/config/run.js:
```diff
@@ -149,6 +149,7 @@ module.exports = function (grunt) {
         '--server.port=' + kibanaTestServerUrlParts.port,
         '--elasticsearch.url=' + esTestConfig.getUrl(),
         '--dev',
+        '--dev_mode.enabled=false',
```
What's this for?

timroes replied:
I actually wonder if any of the devs ever uses that Debug spy panel :D If not we could also get rid of it completely.

chrisronline commented on src/ui/public/notify/toasts/global_toast_list.js:
```diff
@@ -0,0 +1,133 @@
+
+  startDismissingToast(toast) {
+    this.setState(prevState => {
+      const toastIdToDismissedMap = { ...prevState.toastIdToDismissedMap };
```
FYI, this can be written as:

```
const toastIdToDismissedMap = {
   ...prevState.toastIdToDismissedMap,
   [toast.id]: true,
};
```

No need to change, but I sort of like it better this way and wanted to suggest in case you weren't aware.

chrisronline commented on src/ui/public/notify/toasts/global_toast_list.js:
```diff
@@ -0,0 +1,133 @@
+  }
+
+  componentWillUnmount() {
+    this.timeoutIds.forEach(timeoutId => clearTimeout(timeoutId));
```
FYI, this can be written as:

```
this.timeoutIds.forEach(clearTimeout);
```

Again, no need to change, but I sort of like it better this way

timroes commented on src/ui/public/notify/fatal_error.js:
```diff
@@ -0,0 +1,88 @@
+ */
+export function fatalError(err, location) {
+  fatalErrorInternals.show(err, location);
+  console.error(err.stack); // eslint-disable-line no-console
```
Does the `stack` also contain the message? I think we should also log the message to the console, so we don't just have stacktraces, without any related messages in the console, so you would need to make the link between the message shown in a toast and the stacktrace yourself.

cjcenizal replied:
Which message do you mean?

cjcenizal commented:
Thanks @chrisronline and @timroes I've addressed your feedback except for Tim's note about the fatal error.

cjcenizal commented on src/core_plugins/kibana/public/dashboard/dashboard_app.js:
```diff
@@ -183,14 +184,18 @@ app.directive('dashboardApp', function ($injector) {
         dashboardStateManager.addNewPanel(hit.id, 'visualization');
         if (showToast) {
-          notify.info(`Visualization successfully added to your dashboard`);
+          toastNotifications.addSuccess('Added visualization to your dashboard');
```
Per @gchaps: "Added _a_ visualization..."

cjcenizal commented:
I completely removed the `notify.info` method, which means I also had to surface the "No index patterns" message in a different way. CC @nreese 

![image](https://user-images.githubusercontent.com/1238659/35415091-441db448-01d9-11e8-81a9-12289dc190d0.png)

cjcenizal commented:
jenkins test this

cjcenizal commented on src/ui/public/notify/toasts/TOAST_NOTIFICATIONS.md:
```diff
@@ -0,0 +1,79 @@
+# Toast notifications
```
See https://github.com/cjcenizal/kibana/blob/euify/notification-toasts-extract-fatal/src/ui/public/notify/toasts/TOAST_NOTIFICATIONS.md to preview this file in a more readable form.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/15124/)

cjcenizal commented:
jenkins test this

cjcenizal commented:
jenkins, test this

cjcenizal commented:
jenkins test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/15129/)

cjcenizal commented:
jenkins, test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/15133/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/15139/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/15141/)

gchaps commented on src/ui/public/notify/toasts/TOAST_NOTIFICATIONS.md:
````diff
@@ -0,0 +1,100 @@
+Toasts will automatically be dismissed after a brief delay, but if for some reason you want to dismiss a toast, you can use the returned toast from one of the `add` methods and then pass it to `remove`.
+
+```js
+const toast = toastNotifications.add('Saved document');
````
Your document was saved

gchaps commented on src/ui/public/notify/toasts/TOAST_NOTIFICATIONS.md:
````diff
@@ -0,0 +1,100 @@
+
+```js
+toastNotifications.add({
+  title: 'Saved document',
````
Your document was saved

gchaps commented on src/ui/public/notify/toasts/TOAST_NOTIFICATIONS.md:
````diff
@@ -0,0 +1,100 @@
+
+```js
+toastNotifications.add({
+  title: 'Saved document',
````
Your document was saved

Only you have editing privileges.  Edit permisson (no ending period cause its a link)

gchaps commented on src/ui/public/notify/toasts/TOAST_NOTIFICATIONS.md:
````diff
@@ -0,0 +1,100 @@
+
+```js
+toastNotifications.addSuccess({
+  title: 'Saved document',
````
Your document was saved

gchaps commented on src/ui/public/share/directives/share.js:
```diff
@@ -145,23 +145,28 @@ app.directive('share', function (Private) {
           if (isCopied) {
-            notify.info('URL copied to clipboard.');
+            toastNotifications.add({
+              title: 'URL copied to clipboard',
```
URL was copied to the clipboard
